### PR TITLE
Implement SnapStart Runtime Hooks

### PR DESCRIPTION
**Pull request by mbfreder**
*Issue #, if available:*

*Description of changes:*
Implemented SnapStart Runtime Hooks

- Added a crac::Resource trait 
- Implemented Builder pattern to allow registering crac::Resource with the Runtime
- Removed the simulated connector
- Refactor the test code to use httpmock to mock RAPID

By submitting this pull request

- [x] I confirm that my contribution is made under the terms of the Apache 2.0 license.
- [x] I confirm that I've made a best effort attempt to update all relevant documentation.

### CODE REVIEW

File: examples/http-runtime-hooks/README.md [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# AWS Lambda Function example
+
+## Build & Deploy
+
+1. Install [cargo-lambda](https://github.com/cargo-lambda/cargo-lambda#installation)
```
Comment (roger-zhangg) [inline examples/http-runtime-hooks/README.md:5]: Seems https://www.cargo-lambda.info/guide/installation.html makes more sense here. There's no installation secion in this git

File: examples/basic-runtime-hooks/README.md [hunk 2]
```diff
@@ -0,0 +1,11 @@
+# AWS Lambda Function example
+
+## Build & Deploy
+
+1. Install [cargo-lambda](https://github.com/cargo-lambda/cargo-lambda#installation)
```
Comment (roger-zhangg) [inline examples/basic-runtime-hooks/README.md:5]: same as above

File: lambda-runtime/src/requests.rs [hunk 3]
```diff
@@ -255,3 +275,26 @@ fn test_init_error_request() {
...
+    }
+}
+#[test]
+fn test_restore_next_event_request() {
+    let req = RestoreNextRequest;
+    let req = req.into_req().unwrap();
+    assert_eq!(req.method(), Method::GET);
+    assert_eq!(req.uri(), &Uri::from_static("/2018-06-01/runtime/restore/next"));
+    assert!(match req.headers().get("User-Agent") {
+        Some(header) => header.to_str().unwrap().starts_with("aws-lambda-rust/"),
```
Comment (roger-zhangg) [inline lambda-runtime/src/requests.rs:297]: might prefer a Const instead of str value

**Review by roger-zhangg — COMMENTED** [review]
left some comments

**Review by bnusunny — APPROVED** [review]
LGTM, Waiting for a second look from @calavera

**Pull request closed without merging**